Task/user/user controller

petertretyakov opened the pull request:
**Review topic**
#16 

**Review description**
Created `UserController` and `UserUnauthorizedController` for further `UserMiddleware`. Implemented create, update, show methods for user, session generation and session token update after one month, user photo uploads.

Endpoints
`POST /user` — creates User (only `name` and `lastname` fields required)
`GET /user/:id` — returns User (and updates session token if necessary)
`PATCH /user/:id` — updates User (and session token if necessary)

Check UserController tests, maybe I've missed something.

petertretyakov committed:
[user] Updated models

petertretyakov committed:
[user] Created UserController and tests for store and show methods

petertretyakov committed:
[user] Created separate DB configs for development and test environments

petertretyakov committed:
[user] Split User Controller to authorized and unauthorized controllers

petertretyakov committed:
[user] Updated tests for update request

petertretyakov committed:
[user] Added photo upload functionality to update method

petertretyakov committed:
[user] Removed images folder from gitignore

petertretyakov committed:
[user] Updated JSON generation for User and Session models

petertretyakov committed:
[user] Updated test file with Sourcery

petertretyakov committed:
[user] Moved update token functionality from User to Session

petertretyakov committed:
[user] Moved to Session generation through Crypto

ditansu commented on Sources/App/Models/Session/Session.swift:
```diff
@@ -1,8 +1,10 @@
 
 // sourcery: AutoModelGeneratable
-// sourcery: fromJSON, toJSON, Preparation
+// sourcery: Preparation, Timestampable
```
I think you could return toJSON and use `// sourcery: ignoreInJSON`  to `userId` look at `Place` e.g.

dannydaddy3 replied:
@ditansu Yes, he could do this, but I see that there's no "id" also in custom implementation. because we add "id" field in JSON by default in template...

petertretyakov replied:
I'll leave it as is to use it without `id`.

ditansu replied:
Ok

ditansu commented on Tests/AppTests/Controllers/User/UserControllerTests.swift:
```diff
@@ -0,0 +1,180 @@
+  let phone = "1234567890"
+  let photo = "http://cocoaheads.ru/photo.jpg"
+  let updatedLastName = "Feet"
+  
```
You may use  String+Random methods here like String.randomValue, String.randomEmail, randomPhone etc

ditansu commented on Tests/AppTests/Controllers/User/UserControllerTests.swift:
```diff
@@ -0,0 +1,180 @@
+    var json = JSON()
+    try json.set("name", firstName)
+    
+    try drop.clientAuthorizedTestResponse(to: .post, at: "user", body: json).assertStatus(is: .internalServerError)
```
may be change at 
```swift 
try drop
      .clientAuthorizedTestResponse(
        to: .post,
        at: "user",
        body: json)
      .assertStatus(is: .internalServerError)
```

petertretyakov replied:
Fixed

ditansu reviewed (changes requested):
Cool!  but there are some issues IMO)

ditansu commented on Sources/App/Controllers/User/UserController.swift:
```diff
@@ -0,0 +1,77 @@
+  }
+  
+  func newPhotoURL(for userId: String, photoBytes: [UInt8], filename: String) throws -> String {
+    let userDir = URL(fileURLWithPath: droplet.config.workDir)
```
hm.. Vapor have the `Public` folder - it special folder to stuff like images. See https://docs.vapor.codes/2.0/vapor/folder-structure/#minimum-folder-structure

petertretyakov replied:
Fixed

ditansu commented on Sources/App/Controllers/User/UserController.swift:
```diff
@@ -0,0 +1,77 @@
+      let filePathes = try fileManager.contentsOfDirectory(atPath: userDir.path)
+      for path in filePathes {
+        let fullURL = userDir.appendingPathComponent(path)
+        try fileManager.removeItem(at: fullURL)
```
why delete the existed file is so complex? ))

petertretyakov replied:
Discussed

ditansu committed:
Chaned Jenkinsfile, now mysql.conf copy to new target folders.

ditansu committed:
Merge branch 'develop' of https://github.com/cocoaheadsru/server into task/user/user-controller

* 'develop' of https://github.com/cocoaheadsru/server: (26 commits)
  Update README.md
  test2
  fixed fake-broken "test" to pass test-set
  test failed testing stage
  test 2
  don't resolve
  debug
  debug
  debug
  debug
  debug swiftlint 100500
  debug lint 2
  debug lint
  debug swiftlint 3
  debug PR SwiftLint
  debug PR3
  debug PR2
  debug pr
  pr debug
  debug PR check
  ...

# Conflicts:
#	Devops/Jenkinsfile

ditansu committed:
fixed Jenkinsfile formating

ditansu committed:
Fixed minor issues

ditansu committed:
Returned chekstyle config (but it is still commented)

ditansu committed:
Fixed bug with WORK_NODE

ditansu committed:
Fixed issues with some tests

ditansu committed:
Fixed groovy syntax typo

dannydaddy3 commented on Sources/App/Controllers/User/UserController.swift:
```diff
@@ -0,0 +1,77 @@
+    let fileManager = FileManager.default
+    
+    if fileManager.fileExists(atPath: userDir.path) {
+      let filePathes = try fileManager.contentsOfDirectory(atPath: userDir.path)
```
method `contentsOfDirectory` returns not paths, it returns filenames
rename please to `let filenames`

petertretyakov replied:
Fixed

dannydaddy3 commented on Sources/App/Models/Session/Session.swift:
```diff
@@ -11,45 +13,64 @@ final class Session: Model {
+  }
+  
+  func updateToken() throws {
+    if
```
I think this is the case for using `guard`

petertretyakov replied:
You mean `guard` only for `let`s there? Because I also have a condition there `referenceDate < Date()`. So this code runs one time a month in fact and if it can't run it is not an error, except for `let`s initialization.

petertretyakov replied:
```
guard
      let date = updatedAt,
      let referenceDate = Calendar.current.date(byAdding: .month, value: 1, to: date)
      else { throw Abort.serverError }
    if referenceDate < Date() {
      self.token = try Session.generateToken()
      try self.save()
    }
```
Like this?

dannydaddy3 replied:
```
guard
  let date = updatedAt,
  let referenceDate = Calendar.current.date(byAdding: .month, value: 1, to: date),
  referenceDate < Date()
else { throw Abort.serverError }
token = try Session.generateToken()
try self.save()
```
I mean like this, u can add condition in `guard` as well. 
It's just I think, all places where it's only one `if` in function(without `else`), it's a perfect candidate for `guard`

ditansu replied:
@dannydaddy3, `if referenceDate < Date()` you offer throw the server error? I guess it's wrong way.

dannydaddy3 replied:
@ditansu oh, i’m stupid) agree))

dannydaddy3 commented on Sources/App/Models/User/User.swift:
```diff
@@ -35,31 +36,67 @@ final class User: Model {
+    try? json.set(Keys.photo, photo)
+    try? json.set(Keys.email, email)
+    try? json.set(Keys.phone, phone)
+    try json.set("session", session()?.makeJSON())
```
Why code generation didn't work here? It seems no custom stuff here

petertretyakov replied:
I've added it because of additional `session` key. But now it has custom `photo` value. So left it as is.

dannydaddy3 replied:
I got you right, you have only session key that is custom here?
have you tried `// sourcery: nestedJSONRepresentableField` like in `Place` model? 
it will generate session key as well

dannydaddy3 commented on Tests/AppTests/Controllers/User/UserControllerTests.swift:
```diff
@@ -0,0 +1,187 @@
+    }
+  }
+
+  func testThatUserCreatesFromRequest() throws {
```
`testThatUserIsCreatedFromRequest` is more correct i think

petertretyakov replied:
Fixed

dannydaddy3 commented on Tests/AppTests/Controllers/User/UserControllerTests.swift:
```diff
@@ -0,0 +1,187 @@
+    XCTAssertEqual(try User.all().count, 1)
+  }
+  
+  func testThatUserDoesNotCreateFromIncompleteRequest() throws {
```
`testThatUserIsNotCreatedFromIncompleteRequest` is more correct i think

petertretyakov replied:
Fixed

ditansu reviewed (changes requested):
wait changes

petertretyakov committed:
[user] Fixed PR Review 1
+ fixed Danny's EventRegistrationController test with user.photo
+ fixed warning in Dima's Autoapprove test
